File: tests/test_reader.py
import pytest
import sys
import os
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from reader import parse
from Types import value_string


@pytest.mark.parametrize('src', [
    '1 2 | 3',
    'eq 0 | or (item | mod 3 | eq 0)',
])
def test_roundtrip_simple(src):
    ast = parse(src)
    assert parse(value_string(ast)) == ast


def test_roundtrip_example_file():
    with open('examples/euler1.magic') as f:
        src = f.read()
    ast = parse(src)
    assert parse(value_string(ast)) == ast


def test_empty_line_roundtrip():
    src = "1\n\n2"
    ast = parse(src)
    assert len(ast) == 3
    assert isinstance(ast[1], list) and len(ast[1]) == 0
    assert parse(value_string(ast)) == ast


def test_roundtrip_nontrivial_program():
    src = (
        "seq 1 10\n"
        "filter (eq 5)\n"
        "get 0\n"
        "save five\n\n"
        "for (i) (seq 0 five) (\n"
        "echo i\n\n"
        "i | + 2 | echo\n"
        ")"
    )
    ast = parse(src)
    assert parse(value_string(ast)) == ast

File: tests_ruby/test_reader.rb
require 'minitest/autorun'
require_relative '../parse'
require_relative '../types'

class TestReader < Minitest::Test
  def roundtrip(src)
    ast = parse(src)
    assert_equal ast, parse(Types.value_string(ast))
  end

  def test_roundtrip_simple
    ['1 2 | 3', 'eq 0 | or (item | mod 3 | eq 0)'].each do |src|
      roundtrip(src)
    end
  end

  def test_roundtrip_example_file
    src = File.read('examples/euler1.magic')
    roundtrip(src)
  end

  def test_empty_line_roundtrip
    src = "1\n\n2"
    ast = parse(src)
    assert_equal 3, ast.length
    assert ast[1].is_a?(Array) && ast[1].empty?
    assert_equal ast, parse(Types.value_string(ast))
  end

  def test_roundtrip_nontrivial_program
    src = (
      "seq 1 10\n" +
      "filter (eq 5)\n" +
      "get 0\n" +
      "save five\n\n" +
      "for (i) (seq 0 five) (\n" +
      "echo i\n\n" +
      "i | + 2 | echo\n" +
      ")"
    )
    roundtrip(src)
  end
end
